### DISCUSSION TITLE
New %VAR% type that persists in RAM without MEM write overhead.

### DISCUSSION THREAD

**Discussion by GaryMilne** [arendst/Tasmota#15498] [Q&A] [answered]
Hi, Tasmota currently allows storage in either var or mem types. My request is to have some form of variable that persists across rules but does not require writing to NVRAM. What is a use case?

Consider these two rules.
Version [#1](https://github.com/arendst/Tasmota/pull/1)
ON Tele-Energy#Current DO var13 %value% ENDON
ON Tele-Energy#Power DO var14 %value% ENDON "
ON Tele-Energy#Voltage DO backlog ; webquery http://192.168.0.200:39501/ POST {'TSync':'True','Current':'%var13%','Watts':'%var14%','Voltage':'%value%'} ENDON

Basically **sends some JSON to the host at whatever the TELEPERIOD** is, whether the data has changed or not i.e. all zero's. Not ideal from a POV of network usage and host activity and logging.

Version [#2](https://github.com/arendst/Tasmota/issues/2)
ON Tele-Energy#Current DO backlog0 Mem11 %value% ; RuleTimer1 1 ENDON
ON Tele-Energy#Voltage DO backlog0 Mem12 %value% ; RuleTimer1 1 ENDON
ON Tele-Energy#Power DO backlog0 Mem13 %value% ; RuleTimer1 1 ENDON
ON Rules#Timer=1 DO Mem15 %Mem10%,%Mem11%,%Mem12%,%Mem13% ENDON
ON Mem15#State$!%Mem16% DO backlog ; Mem16 %Mem15% ; webquery http://192.168.0.200:39501/ POST {'TasmotaSync':'True','%Mem10%','Current':'%Mem11%','Voltage':'%Mem12%','Watts':'%Mem13%'} ENDON

In this case the net result is that Tasmota **only sends changes to the host whenever one of the three values has changed since the prior period.** Slightly more processing but greatly advantageous in terms of reducing network traffic and events at the host because only changes are sent. However, the considerable downside is that with a TelePeriod of 60 seconds you would have 60 (events) x 24 (hours) x 5 (mem variables) = **7,200 mem writes per day**. That would start killing off memory cells within a few days and obviously a non-starter.

The availability of a couple of semi-persistent (until restart) in memory registers %pvar% would greatly enhance what is capable of achieving with rules within the stock compilation.

Thanks for listening.

**Comment by barbudor** [comment, marked as answer]
Nothing prevent you to use %var% instead of %mem% in your second use-case

%var% are persistant until next restart. So the only drawback to using %var% is to possibly, resend the same data right after a restart which should not be a problem as this should not happen often.

RAM is fully set to 0 at restart. So it is not possible to persist value in RAM.

**Comment by GaryMilne** [comment]
Thank you for clarifying that for me.  The documentation said this about
vars "All var will be empty strings when the program starts."  I
misinterpreted that to mean that they would be blank at the beginning of
each rule execution.  I'm glad to be wrong about this.  I converted my
rules over to var and I'm now getting the best of both worlds.

On Sat, Apr 30, 2022 at 2:49 PM Barbudor ***@***.***> wrote:

> Nothing prevent you to use %var% instead of %mem% in your second use-case
>
> %var% are persistant until next restart. So the only drawback to using
> %var% is to possibly, resend the same data right after a restart which
> should not be a problem as this should not happen often.
>
> RAM is fully set to 0 at restart. So it is not possible to persist value
> in RAM.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/15498#discussioncomment-2667206>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4ZMVLHKZFOXCSNACSMLCLVHWFD3ANCNFSM5UYXE23A>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Gary Milne
E-Mail: ***@***.***
